File: 1_Android/asgn2/Assignment2/app/src/main/java/edu/ucsc/cse118/assignment2/DataClasses.kt
package edu.ucsc.cse118.assignment2

class DataClasses {
    data class Workspace(
        val name: String,
        val channels: List<Channel>
    )

    data class Channel(
        val name: String,
        val messages: List<Message>
    )

    data class Message(
        val user: User,
        val date: String,
        val content: String
    )

    data class User(
        val name: String,
        val email: String
    )

}

File: 3_ReactNative/Assignment8/src/Model/DataClasses.js
export class Workspace {
  constructor(id, name, channels) {
    this.id = id;
    this.name = name;
    this.channels = channels;
  }

  get uniquePosters() {
    const members = this.channels.flatMap(channel =>
      channel.messages.map(message => message.member.id),
    );
    return new Set(members).size;
  }

  get mostRecentMessage() {
    return Math.max(...this.channels.map(channel => channel.mostRecentMessage));
  }
}

export class Channel {
  constructor(id, name, messages) {
    this.id = id;
    this.name = name;
    this.messages = messages;
  }

  get uniquePosters() {
    const members = this.messages.map(message => message.member.id);
    return new Set(members).size;
  }

  get mostRecentMessage() {
    return Math.max(...this.messages.map(message => message.posted));
  }
}

export class Message {
  constructor(id, content, posted, member) {
    this.id = id;
    this.content = content;
    this.posted = new Date(posted);
    this.member = member;
  }
}

export class Member {
  constructor(id, name) {
    this.id = id;
    this.name = name;
  }
}
